lib/web/command.go

```diff
@@ -20,20 +20,611 @@
 package web
 
 import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"net"
 	"net/http"
+	"strconv"
+	"strings"
+	"time"
 
+	"github.com/gogo/protobuf/proto"
+	"github.com/gorilla/websocket"
 	"github.com/gravitational/trace"
 	"github.com/julienschmidt/httprouter"
+	"github.com/sirupsen/logrus"
+	oteltrace "go.opentelemetry.io/otel/trace"
+	"golang.org/x/crypto/ssh"
+	"google.golang.org/protobuf/types/known/timestamppb"
 
+	"github.com/gravitational/teleport"
+	authproto "github.com/gravitational/teleport/api/client/proto"
+	apidefaults "github.com/gravitational/teleport/api/defaults"
+	"github.com/gravitational/teleport/api/gen/proto/go/assist/v1"
+	"github.com/gravitational/teleport/api/observability/tracing"
+	"github.com/gravitational/teleport/lib/agentless"
+	"github.com/gravitational/teleport/lib/client"
+	"github.com/gravitational/teleport/lib/defaults"
+	"github.com/gravitational/teleport/lib/httplib"
+	"github.com/gravitational/teleport/lib/modules"
+	"github.com/gravitational/teleport/lib/proxy"
 	"github.com/gravitational/teleport/lib/reversetunnel"
+	"github.com/gravitational/teleport/lib/session"
+	"github.com/gravitational/teleport/lib/teleagent"
+	"github.com/gravitational/teleport/lib/utils"
 )
 
+// CommandResultType is the type of Assist message that contains the command execution result.
+const CommandResultType = "COMMAND_RESULT"
+
+type CommandRequest struct {
+	// Command is the command to be executed on all nodes.
+	Command string `json:"command"`
+	// Login is a Linux username to connect as.
+	Login string `json:"login"`
+	// Query is the predicate query to filter nodes where the command will be executed.
+	Query string `json:"query"`
+	// ConversationID is the conversation context that was used to execute the command.
+	ConversationID string `json:"conversation_id"`
+	// ExecutionID is a unique ID used to identify the command execution.
+	ExecutionID string `json:"execution_id"`
+}
+
+func (c *CommandRequest) Check() error {
```

Comment:
Can query be empty? It's not validated here.

Reply:
The idea behind it was that an empty query returns all nodes, so the client passing an empty query wants to execute a command on all nodes (that they have access to). I'm not aware of how you can create a query asking to do that, so that was kind of a workaround, but the more I think about it more I lean to block this edge case, as someone may, by mistake execute a command on all nodes.